### PR TITLE
Refactor and test

### LINKED ISSUE #12: Refactor into separate modules and test

### PR DESCRIPTION
**Pull request by simonplend**
Fixes #12.

### CODE REVIEW

File: test/commands/add.test.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+test('running command handler with valid options generates expected log messages', async () => {
+	await addCommand.handler({
+		repo: 'https://github.com/financial-times-sandbox/Timely-Moving-Coffin',
+		config: fixtures.paths.validConfig,
+		token: '123abc'
+	});
+
+	expect(generateLoggerSnapshot()).toMatchInlineSnapshot(
+		`"{\\"info\\":[[\\"Config: Read from local file: /home/simonplend/dev/work/clients/financial-times/managing-repos/manage-github-apps/test/commands/fixtures/config.json\\\\n\\"],[\\"The options you have specified have been parsed as:\\"],[\\"- GitHub organisation: financial-times-sandbox\\"],[\\"- GitHub repo: Timely-Moving-Coffin\\\\n\\"]],\\"message\\":[],\\"success\\":[[\\"Authenticated as GitHub user testuser\\"],[\\"GitHub repo financial-times-sandbox/Timely-Moving-Coffin exists\\\\n\\"]],\\"warning\\":[],\\"error\\":[],\\"custom\\":[[\\"➕\\",\\"Adding repo to installation Some bot (https://github.com/organizations/financial-times-sandbox/settings/installations/123456)\\"],[\\"\\\\n➡️\\",\\"Go to https://github.com/financial-times-sandbox/Timely-Moving-Coffin/settings/installations to see the installed GitHub apps for this repo.\\"]]}"`
+	);
```
Comment (simonplend) [inline test/commands/add.test.js]: I opted for a snapshot here, as the alternative was:

```
const infoCalls = logger.info.mock.calls;
expect(infoCalls[0][0]).toMatch(new RegExp(`^Config: Read from local file:.+${escapeRegExp(fixtures.paths.validConfig)}`));
expect(infoCalls[1][0]).toMatch('The options you have specified have been parsed as:');
expect(infoCalls[2][0]).toMatch('- GitHub organisation: financial-times-sandbox');
expect(infoCalls[3][0]).toMatch('- GitHub repo: Timely-Moving-Coffin');
```

Thoughts?

File: test/commands/add.test.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		command: expect.stringMatching('add'),
+		desc: expect.any(String),
+		builder: expect.any(Function),
+		handler: expect.any(Function)
+	});
+});
+
+test('yargs can load the `add` command without any warnings', () => {
+	yargs.command(addCommand.command, addCommand.desc, addCommand.builder, addCommand.handler).argv;
+	expect(mockConsoleWarn).toHaveBeenCalledTimes(0);
```
Comment (sjparkinson) [inline test/commands/add.test.js]: Would it make sense to add a no throw expectation here too?

File: test/commands/add.test.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	mockConsoleWarn.mockRestore();
+});
+
+test('`add` command module exports an object that can be used by yargs', () => {
+	expect.objectContaining({
+		command: expect.stringMatching('add'),
+		desc: expect.any(String),
+		builder: expect.any(Function),
+		handler: expect.any(Function)
+	});
```
Comment (sjparkinson) [inline test/commands/add.test.js]: Awesome 💯

File: test/commands/add.test.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	expect(mockConsoleWarn).toHaveBeenCalledTimes(0);
+});
+
+test('running command handler without `repo` will exit process with error', async () => {
+	await addCommand.handler({
+		config: fixtures.paths.validConfig,
+		token: '123abc'
+	});
+	expect(logger.error).toBeCalledWith(
+		'ERROR: Github#extractOwnerAndRepo: Could not extract owner and repo from provided string'
```
Comment (sjparkinson) [inline test/commands/add.test.js]: Are we sure we want to couple these expectations to the actual message?

Perhaps an error code that shouldn't change would make more sense. We've often used `event: ERROR_SOME_UNIQUE_ERROR`, but also nice is Heroku's `error=H10` style.

File: test/commands/add.test.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+// User modules
+jest.mock('../../src/lib/logger');
+const logger = require('../../src/lib/logger');
+
+const addCommand = require('../../src/commands/add');
+
+const fixtures = {
+	paths: {
+		validConfig: 'test/commands/fixtures/config.json'
+	}
```
Comment (adambraimbridge) [inline test/commands/add.test.js]: I put fixtures into the __mocks__ directory. Which is probably wrong :)

File: test/commands/add.test.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+const generateLoggerSnapshot = () => {
```
Comment (adambraimbridge) [inline test/commands/add.test.js]: Does jest not generate snapshots automagically?

Comment (adambraimbridge) [reply test/commands/add.test.js]: Oh, I guess, not for logs, like this case.

File: test/commands/add.test.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+beforeEach(() => {
+	mockProcessExit = jest
+		.spyOn(process, 'exit')
+		.mockImplementation((code) => code);
+
+	mockConsoleWarn = jest.spyOn(console, 'warn');
+});
+
+afterEach(() => {
+	// TODO: What kind of clearing needs to happen here? Do we need to reset spies?
```
Comment (adambraimbridge) [inline test/commands/add.test.js]: Yeah, I think I recall needing to reset spies for each test.

File: test/commands/add.test.js (hunk 2 above)
Comment (adambraimbridge) [inline test/commands/add.test.js]: FWIW, I think you can have `.not.toHaveBeenCalled()` ... But of course that's not necessarily better.

File: test/commands/add.test.js (hunk 4 above)
Comment (adambraimbridge) [inline test/commands/add.test.js]: Personally, I'd be less explicit. Maybe something like this:

`expect(logger.error).toBeCalledWith(expect.stringContaining('ERROR: Github#extractOwnerAndRepo'))`

File: test/commands/add.test.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+});
+
+test('running command handler with valid options generates expected log messages', async () => {
+	await addCommand.handler({
+		repo: 'https://github.com/financial-times-sandbox/Timely-Moving-Coffin',
+		config: fixtures.paths.validConfig,
+		token: '123abc'
+	});
+
+	expect(generateLoggerSnapshot()).toMatchInlineSnapshot(
```
Comment (adambraimbridge) [inline test/commands/add.test.js]: Whoa, this is cool

File: test/commands/add.test.js (hunk 1 above)
Comment (adambraimbridge) [reply test/commands/add.test.js]: Whoa, this is cool

Comment (adambraimbridge) [reply test/commands/add.test.js]: I'm curious. Could you not call `.toMatchSnapshot()` here?

File: package.json [hunk 9]
```diff
@@ -7,9 +7,10 @@
     "mga": "src/index.js"
   },
   "scripts": {
+    "test": "npm run lint && npm run unit-test",
     "lint": "eslint --fix src/",
```
Comment (adambraimbridge) [inline package.json]: Unrelated: I don't like having `--fix` applied here :(

File: test/commands/add.test.js (hunk 5 above)
Comment (simonplend) [reply test/commands/add.test.js]: Ahh! As the timeless saying goes, "When is a fixture not a mock? When it's a fixture!"

File: test/commands/add.test.js (hunk 6 above)
Comment (simonplend) [reply test/commands/add.test.js]: Yeah, I'll give this a better name - it's not actually generating the Jest snapshot, but the contents of it.

File: test/commands/add.test.js (hunk 7 above)
Comment (simonplend) [reply test/commands/add.test.js]: Makes sense :+1:

File: test/commands/add.test.js (hunk 3 above)
Comment (simonplend) [reply test/commands/add.test.js]: :blush:

File: test/commands/add.test.js (hunk 2 above)
Comment (simonplend) [reply test/commands/add.test.js]: Of _course_. That is precisely what i wanted to do. Thanks! :man_facepalming:

Comment (simonplend) [reply test/commands/add.test.js]: Might well do, I'll look into it. At the very least I'll put a comment here that `yargs` just uses `console.warn` to raise errors about incorrect types for arguments to the `command` method (which I find kind of weird to be honest).

File: test/commands/add.test.js (hunk 4 above)
Comment (simonplend) [reply test/commands/add.test.js]: Indeed - I totally dislike what I've done here. I think I'll take on Adam's suggestion below though.

File: test/commands/add.test.js (hunk 1 above)
Comment (simonplend) [reply test/commands/add.test.js]: Yep, the snapshot could definitely be stored in a snapshot file, I just chose to inline it as part of my learning exercise of using Jest snapshots for the first time. Do you see pros / cons either way?

Comment (simonplend) [reply test/commands/add.test.js]: I've ended up putting the snapshot in a file as it allows for it to be nicely formatted and keeps the tests short.

File: test/commands/add.test.js (hunk 5 above)
Comment (adambraimbridge) [reply test/commands/add.test.js]: 😂

**Pull request merged**